code/modules/mob/living/simple_animal/hostile/goose.dm

```diff
@@ -25,33 +27,144 @@
+	if(!istype(O, /obj/item/reagent_containers/food))
+		return
+	if (contents.len > GOOSE_SATIATED)
+		visible_message("<span class='notice'>[src] looks too full to eat \the [tasty]!</span>")
+		return
+	if (tasty.foodtype & GROSS)
+		visible_message("<span class='notice'>[src] hungrily gobbles up \the [tasty]!</span>")
+		tasty.forceMove(src)
+		playsound(src,'sound/items/eatfood.ogg', 70, 1)
+		vomitCoefficient += 3
+		vomitTimeBonus += 2
 	else
-		icon_state = icon_dead
+		visible_message("<span class='notice'>[src] refuses to eat \the [tasty].</span>")
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/vomit()
+	var/turf/T = get_turf(src)
+	var/obj/item/reagent_containers/food/consumed = locate() in contents //Barf out a single food item from our guts
+	if (prob(50) && consumed)
+		barf_food(consumed)
+	else
+		playsound(get_turf(src), 'sound/effects/splat.ogg', 50, 1) //yes getting the turf twice is necessary fuck off
+		T.add_vomit_floor(src)
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/barf_food(var/atom/A, var/hard = FALSE)
+	if(!istype(A, /obj/item/reagent_containers/food))
+		return
+	var/turf/currentTurf = get_turf(src)
+	var/obj/item/reagent_containers/food/consumed = A
+	consumed.forceMove(currentTurf)
+	var/destination = get_edge_target_turf(currentTurf, pick(GLOB.alldirs)) //Pick a random direction to toss them in
+	var/throwRange = hard ? rand(2,8) : 1
+	consumed.safe_throw_at(destination, throwRange, 2) //Thow the food at a random tile 1 spot away
+	sleep(2)
+	if (QDELETED(src) || QDELETED(consumed))
+		return
+	currentTurf = get_turf(consumed)
+	currentTurf.add_vomit_floor(src)
+	playsound(get_turf(consumed), 'sound/effects/splat.ogg', 50, 1) //yes getting the turf twice is necessary fuck off
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/vomit_prestart(duration)
+	flick("vomit_start",src)
+	addtimer(CALLBACK(src, .proc/vomit_start, duration), 13) //13 is the length of the vomit_start animation in gooseloose.dmi
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/vomit_start(duration)
+	vomiting = TRUE
+	icon_state = "vomit"
+	vomit()
+	addtimer(CALLBACK(src, .proc/vomit_preend), duration)
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/vomit_preend()
+	for (var/obj/item/consumed in contents) //Get rid of any food left in the poor thing
+		barf_food(consumed, TRUE)
+		sleep(1)
+		if (QDELETED(src))
+			return
+	vomit_end()
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/proc/vomit_end()
+	flick("vomit_end",src)
+	vomiting = FALSE
+	icon_state = initial(icon_state)
+
+/mob/living/simple_animal/hostile/retaliate/goose/vomit/Moved(oldLoc, dir)
+	. = ..()
+	if(vomiting)
+		vomit() // its supposed to keep vomiting if you move
+		return
+	var/turf/currentTurf = get_turf(src)
+	for (var/obj/item/reagent_containers/food/tasty in currentTurf)
+		if (tasty && get_turf(tasty) == currentTurf) //Make sure the food hasn't moved from our turf
+			feed(tasty)
```

Comment:
Because you're sleeping the food may have moved by this point, you need to use a loop that each iteration looks for and grabs a food on the tile if there are any

Reply:
< clueless

Reply:
@ninjanomnom Im pretty sure that doesn't happen

Reply:
https://i.imgur.com/phzNp40.gifv

Reply:
I'm talking about how the loop works, I know you have some checks for this but you shouldn't need those checks if you used a better style of loop.

Reply:
@ninjanomnom Tried this, goes way too slowly for the memes this goose is meant for.

Reply:
If the processing is to slow for your taste then yes you need some sort of loop like this. It still needs to be changed though to not be a for loop through turf contents and instead grab a new food item each loop

Reply:
@ninjanomnom So a recursive function?

Reply:
Did you read the example code I put on discord? That's basically what I suggested

Reply:
@ninjanomnom I didn't see your discord message, but alright, I can add the while loop thing then